// @cloudbase/model/dist/cjs/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerModel = exports.lazyGetEntity = void 0;
var wx_cloud_client_sdk_1 = require("@cloudbase/wx-cloud-client-sdk");
var utilities_1 = require("@cloudbase/utilities");
var CloudbaseEventEmitter = utilities_1.events.CloudbaseEventEmitter;
var COMPONENT_NAME = 'models';
function getEntity(cloudbase) {
    var _this = this;
    var callFunction = cloudbase.callFunction.bind(cloudbase);
    var cloudbaseFetch = cloudbase.request.fetch.bind(cloudbase.request);
    var fetch = function (fetchOptions) { return __awaiter(_this, void 0, void 0, function () {
        var res;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4, cloudbaseFetch(__assign(__assign({}, fetchOptions), { headers: __assign({ 'Content-Type': 'application/json' }, fetchOptions.headers) }))];
                case 1:
                    res = _a.sent();
                    return [4, res.data];
                case 2: return [2, _a.sent()];
            }
        });
    }); };
    var _a = cloudbase.getEndPointWithKey('GATEWAY'), BASE_URL = _a.BASE_URL, PROTOCOL = _a.PROTOCOL;
    var baseUrl = "".concat(PROTOCOL).concat(BASE_URL, "/model");
    var sqlBaseUrl = "".concat(PROTOCOL).concat(BASE_URL, "/sql");
    var entity = (0, wx_cloud_client_sdk_1.generateHTTPClient)(callFunction, fetch, baseUrl, { sqlBaseUrl: sqlBaseUrl });
    return entity;
}
var cloudbaseModelMap = new WeakMap();
function lazyGetEntity(cloudbase) {
    return new Proxy({}, {
        get: function (_, prop) {
            var entity = cloudbaseModelMap.get(cloudbase);
            if (!entity) {
                entity = getEntity(cloudbase);
                cloudbaseModelMap.set(cloudbase, entity);
            }
            return entity[prop];
        },
    });
}
exports.lazyGetEntity = lazyGetEntity;
var CLOUDBASE_INIT_EVENT = 'cloudbase_init';
var bus = new CloudbaseEventEmitter();
bus.on(CLOUDBASE_INIT_EVENT, function (_a) {
    var cloudbase = _a.data;
    Object.assign(cloudbase, { models: lazyGetEntity(cloudbase) });
});
var component = {
    name: COMPONENT_NAME,
    namespace: COMPONENT_NAME,
    entity: new Proxy({}, {
        get: function (_, prop) {
            console.warn("\u3010deprecated\u3011Accessing Cloudbase.prototype.models.".concat(prop, "."));
        },
    }),
    injectEvents: {
        bus: bus,
        events: [CLOUDBASE_INIT_EVENT],
    },
};
try {
    cloudbase.registerComponent(component);
}
catch (e) { }
function registerModel(app) {
    try {
        app.registerComponent(component);
    }
    catch (e) {
        console.warn(e);
    }
}
exports.registerModel = registerModel;
__exportStar(require("@cloudbase/wx-cloud-client-sdk"), exports);
